src/scanner.hpp

```diff
@@ -4,17 +4,17 @@
 #include <cstddef>
 #include <string>
 #include <string_view>
-#include <unordered_map>
+#include <utility>
 #include <vector>
 
 class Scanner
 {
 public:
   explicit Scanner(std::string_view source);
   [[nodiscard]] std::vector<Token> const &get_tokens() const { return tokens; }
+  [[nodiscard]] std::pair<bool, std::vector<Token>> scan();
```

Comment:
Don't use `std::pair`, name your types and members meaningfully. Besides, what does the bool even represent? I mean what is the caller expected to do based on it being true/false?

Reply:
The `bool` represents the `error` state. It replaces the previously defined `static bool error` in the `Clox` class.

Reply:
Is `std::vector<Token>` being empty valid? If not, that itself can be the signal that an error occurred. If yes, you can still return `std::optional<std::vector<Token>>` and use `std::nullopt` as the error signal. Neither of these is an option if you need the tokens scanned even if there was an error. In that case create a meaningfully named struct (and members).